Add "no-identical-names" rule

Turbo87 opened the pull request:
This PR implements a `no-identical-names` rule similar to https://github.com/lo1tuma/eslint-plugin-mocha/blob/master/docs/rules/no-identical-title.md but for QUnit.

Resolves #56 

/cc @platinumazure

platinumazure commented:
@Turbo87 Thanks for writing a PR! Unfortunately Travis didn't run due to the S3 issues at Amazon. Allegedly those are back up again.

Your commit message also has a typo in it, so maybe you could `git commit --amend` and then `git push origin no-identical-title -f` to kill two birds with one stone? :smile: Thanks!

platinumazure commented on docs/rules/no-identical-title.md:
```diff
@@ -0,0 +1,48 @@
+# Forbid identical test and test suite titles (no-identical-title)
+
+Having identical titles for two different tests or test suites may create
```
Let's use QUnit's terminology: modules and tests. Thanks!

platinumazure commented on docs/rules/no-identical-title.md:
```diff
@@ -0,0 +1,48 @@
+
+Having identical titles for two different tests or test suites may create
+confusion. For example, when a test with the same title as another test
+in the same test suite fails, it is harder to know which one failed and
```
s/test suite/module

platinumazure commented on docs/rules/no-identical-title.md:
```diff
@@ -0,0 +1,48 @@
+
+## Further Reading
+
+* [QUnit.module()](http://api.qunitjs.com/QUnit.module/)
```
Might be worth adding a link to `QUnit.test()` as well?

platinumazure commented on lib/rules/no-identical-title.js:
```diff
@@ -0,0 +1,72 @@
+"use strict";
```
Please add a JSDoc fileoverview comment (see https://github.com/platinumazure/eslint-plugin-qunit/blob/master/lib/rules/literal-compare-order.js#L1-L4). You can even put your name down for author since you created this rule :smile:

platinumazure commented on lib/rules/no-identical-title.js:
```diff
@@ -0,0 +1,72 @@
+}
+
+
+function handlTestCaseTitles(context, titles, node, title) {
```
This should probably say `handleTestTitles`.

platinumazure commented on lib/rules/no-identical-title.js:
```diff
@@ -0,0 +1,72 @@
+    }
+}
+
+function handlTestSuiteTitles(context, titles, node, title) {
```
This should probably say `handleModuleTitles`.

platinumazure commented on lib/rules/no-identical-title.js:
```diff
@@ -0,0 +1,72 @@
+    if (titles.indexOf(title) !== -1) {
+        context.report({
+            node: node.arguments[0],
+            message: "Test suite title is used multiple times."
```
Nitpick, but I'd love to see this say "Module name" instead of "Test suite title".

platinumazure commented on lib/rules/no-identical-title.js:
```diff
@@ -0,0 +1,72 @@
+        if (titles.indexOf(title) !== -1) {
+            context.report({
+                node: node.arguments[0],
+                message: "Test title is used multiple times in the same test module."
```
Nitpick, but I'd love to see this say "Test name" instead of "Test title", as well as just "module" instead of "test module".

Turbo87 commented:
@platinumazure thanks for the quick review. I think I've addressed all your comments now. I've also renamed the rule to `no-identical-names` since QUnit lingo seems to prefer `name` over `title`.

platinumazure commented on tests/lib/rules/no-identical-names.js:
```diff
@@ -0,0 +1,111 @@
+                "test(`it${n}`, function() {});"
+            ].join("\n"),
+            env: {
+                es6: true
```
I think I'd prefer `parserOptions: { ecmaVersion: 6 }` here-- setting the es6 environment also enables some globals and other things and I'd rather just focus on the parsing capability of allowing template strings. Sorry for not catching this the first time around.

platinumazure commented on tests/lib/rules/no-identical-names.js:
```diff
@@ -0,0 +1,111 @@
+                column: 8,
+                line: 3
+            }]
+        }
```
General note: Could we have a few tests using modules with function expressions (without nesting)?

For example, here's a valid case:

```js
QUnit.module("module1", function() {
    QUnit.test("test1", function() {});
    QUnit.test("test2", function() {});
});
```

I'm fairly sure they'll work, but it will be good to know with certainty. :smile:

Turbo87 commented:
@platinumazure adjusted, thanks!

platinumazure commented:
@Turbo87 Thanks very much! This provisionally looks good to me. I just want to try your branch against my work codebase (which I've found very handy for exposing unintentional bugs or crashes due to the outright *pathological* code styles some of my coworkers have adopted), and if all goes well there, I'll merge, hopefully by the end of the week. Thanks for your patience and for responding so rapidly to feedback!

Turbo87 commented:
btw I've enabled the "Allow edits from maintainers" checkbox, so feel free to add failing test cases if you find something in your codebase that is not behaving as expected.

platinumazure commented:
Happy to report that the rule didn't crash and, as a bonus, it even identified identical test names correctly. I'll be cleaning up 203 tests once I consume this rule... Thanks very much for contributing, this is awesome and I'll cut a minor release this weekend!

Turbo87 commented:
Do you think it would make sense to add the rule to one of the default rulesets?

platinumazure commented:
I could see an argument for adding it to qunit/default with the next major
release. I can create an issue (or you can if you beat me to it, I guess).

On Mar 2, 2017 2:14 AM, "Tobias Bieniek" <notifications@github.com> wrote:

> Do you think it would make sense to add the rule to one of the default
> rulesets?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/platinumazure/eslint-plugin-qunit/pull/57#issuecomment-283585664>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARWenxuIA-w06tw7TTiAImaEzut0e2uks5rhno4gaJpZM4MO6bw>
> .
>

Turbo87 commented:
@platinumazure any news on the release?

platinumazure commented:
@Turbo87 Apologies for the delay-- meant to release this past weekend but a bunch of things came up. It is now released as `eslint-plugin-qunit@2.4.0`. Thanks again for your contribution!
